File: packages/ts/core/src/telemetry.ts
import type { TokenUsage } from "./models/index.js";

/** Identifying context passed to every telemetry callback. */
export interface TelemetryContext {
  traceId: string;
  providerId: string;
  method: string;
}

/**
 * The observability seam the orchestrator drives. Every callback is optional so
 * an implementation opts into only the events it cares about; the orchestrator
 * guards each call with `?.`. The lifecycle per call is
 * `onRequestStart` → (`onToken`…)? → `onRequestEnd` on success, or
 * `onRequestStart` → `onError` on failure.
 */
export interface TelemetryHook {
  onRequestStart?(ctx: TelemetryContext): void;
  onRequestEnd?(ctx: TelemetryContext, usage: TokenUsage): void;
  onError?(ctx: TelemetryContext, err: unknown): void;
  onToken?(ctx: TelemetryContext, token: string): void;
}

/**
 * Default hook: telemetry is opt-in, so the absence of every callback makes all
 * events no-ops (the orchestrator calls them with `?.`). Never throws, never
 * emits output.
 */
export const NoopTelemetryHook: TelemetryHook = {};

/**
 * Emits one structured record per event, carrying the `traceId`, `providerId`,
 * and method, plus the normalized {@link TokenUsage} on `request.end`. The sink
 * defaults to `console.log(JSON.stringify(...))` but is injectable. Each method
 * swallows sink errors so telemetry can never break the request path.
 */
export class StructuredLogTelemetryHook implements TelemetryHook {
  constructor(
    private readonly log: (rec: object) => void = (r) =>
      console.log(JSON.stringify(r)),
  ) {}

  private emit(rec: object): void {
    try {
      this.log(rec);
    } catch {
      // Telemetry must never surface its own failure into the request path.
    }
  }

  onRequestStart(ctx: TelemetryContext): void {
    this.emit({ event: "request.start", ...ctx });
  }

  onRequestEnd(ctx: TelemetryContext, usage: TokenUsage): void {
    this.emit({ event: "request.end", ...ctx, usage });
  }

  onError(ctx: TelemetryContext, err: unknown): void {
    this.emit({
      event: "error",
      ...ctx,
      error: err instanceof Error ? err.message : String(err),
    });
  }

  onToken(ctx: TelemetryContext, token: string): void {
    this.emit({ event: "token", ...ctx, token });
  }
}

File: packages/py/llmorch-core/src/llmorch_core/telemetry.py
"""Telemetry hooks — twin of packages/ts/core/src/telemetry.ts.

A ``TelemetryHook`` Protocol with four callbacks the orchestrator drives:
``on_request_start`` before the (resilient) provider call, ``on_request_end``
with the response's normalized ``TokenUsage`` on success, ``on_error`` on
failure, and ``on_token`` per chunk while streaming. Ships a no-op default
(opt-in telemetry) and a structured-logging implementation.
"""

import json
from collections.abc import Callable
from dataclasses import asdict, dataclass
from typing import Any, Protocol, runtime_checkable

from .models import TokenUsage


@dataclass
class TelemetryContext:
    """Identifying context passed to every telemetry callback."""

    trace_id: str
    provider_id: str
    method: str


@runtime_checkable
class TelemetryHook(Protocol):
    """The observability seam the orchestrator drives (lifecycle per call:
    ``on_request_start`` → (``on_token``…)? → ``on_request_end`` on success, or
    ``on_request_start`` → ``on_error`` on failure)."""

    def on_request_start(self, ctx: TelemetryContext) -> None: ...
    def on_request_end(self, ctx: TelemetryContext, usage: TokenUsage) -> None: ...
    def on_error(self, ctx: TelemetryContext, err: BaseException) -> None: ...
    def on_token(self, ctx: TelemetryContext, token: str) -> None: ...


class NoopTelemetryHook:
    """Default hook: telemetry is opt-in, so every callback is a safe no-op."""

    def on_request_start(self, ctx: TelemetryContext) -> None:
        pass

    def on_request_end(self, ctx: TelemetryContext, usage: TokenUsage) -> None:
        pass

    def on_error(self, ctx: TelemetryContext, err: BaseException) -> None:
        pass

    def on_token(self, ctx: TelemetryContext, token: str) -> None:
        pass


class StructuredLogTelemetryHook:
    """Emits one structured record per event, carrying the trace/provider/method
    and the normalized ``TokenUsage`` on ``request.end``. The sink defaults to
    ``print(json.dumps(...))`` but is injectable; each method swallows sink
    errors so telemetry never breaks the request path."""

    def __init__(self, log: Callable[[dict[str, Any]], None] | None = None) -> None:
        self._log = log if log is not None else (lambda r: print(json.dumps(r)))

    def _emit(self, rec: dict[str, Any]) -> None:
        try:
            self._log(rec)
        except Exception:  # noqa: BLE001 — telemetry must never surface its own failure
            pass

    def on_request_start(self, ctx: TelemetryContext) -> None:
        self._emit({"event": "request.start", **asdict(ctx)})

    def on_request_end(self, ctx: TelemetryContext, usage: TokenUsage) -> None:
        self._emit(
            {"event": "request.end", **asdict(ctx), "usage": usage.model_dump(by_alias=True)}
        )

    def on_error(self, ctx: TelemetryContext, err: BaseException) -> None:
        self._emit({"event": "error", **asdict(ctx), "error": str(err)})

    def on_token(self, ctx: TelemetryContext, token: str) -> None:
        self._emit({"event": "token", **asdict(ctx), "token": token})
